File: Lancement d'un programme exec/exo6.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

int main(int argc, char *argv[]) 
{
   
    int n = atoi(argv[1]);
    int k = atoi(argv[2]);

    
    if (k == 0 || k == n) 
    {
        printf("1\n");
        return 1; 
    }
    char n1[10], k1[10], k2[10], fname[50];
    sprintf(n1, "%d", n - 1);
    sprintf(k1, "%d", k - 1);
    sprintf(k2, "%d", k);
    sprintf(fname, "/tmp/cnk_%d_%d", n, k);

    
    if (fork() == 0) 
    {
        execl(argv[0], argv[0], n1, k1, NULL);
    }
    if (fork() == 0) 
    {
        execl(argv[0], argv[0], n1, k2, NULL);
    }
    int s1, s2;
    wait(&s1);
    wait(&s2);

    
    int resultat = WEXITSTATUS(s1) + WEXITSTATUS(s2); 
    
    if (argc == 3) 
    {
        printf("Resultat C(%d, %d) = %d\n", n, k, resultat);
    }
    return resultat;
}

File: Lancement d'un programme exec/exo6.py
import os
import sys

def main():
    # Récupération des arguments n et k
    n = int(sys.argv[1])
    k = int(sys.argv[2])

    # Cas de base
    if k == 0 or k == n:
        # En Python, on utilise sys.exit() pour renvoyer un code de retour
        sys.exit(1)

    # Préparation des arguments pour les appels récursifs
    n1 = str(n - 1)
    k1 = str(k - 1)
    k2 = str(k)

    # Premier fils : calcul de C(n-1, k-1)
    pid1 = os.fork()
    if pid1 == 0:
        os.execl(sys.executable, sys.executable, sys.argv[0], n1, k1)
        os._exit(0)

    # Deuxième fils : calcul de C(n-1, k)
    pid2 = os.fork()
    if pid2 == 0:
        os.execl(sys.executable, sys.executable, sys.argv[0], n1, k2)
        os._exit(0)


    _, s1 = os.wait()
    _, s2 = os.wait()


    res1 = os.waitstatus_to_exitcode(s1)
    res2 = os.waitstatus_to_exitcode(s2)

    resultat = res1 + res2

    
    if len(sys.argv) == 3:
        print(f"Resultat C({n}, {k}) = {resultat}")

    sys.exit(resultat)

if __name__ == "__main__":
    if len(sys.argv) < 3:
        sys.exit(1)
    main()